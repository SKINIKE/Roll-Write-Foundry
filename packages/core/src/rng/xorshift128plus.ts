import { SerializedRngState } from '../types';

export type SeedInput = number | bigint | string | [bigint, bigint] | SerializedRngState;

const UINT64_MASK = (1n << 64n) - 1n;

function toUint64(value: bigint): bigint {
  return value & UINT64_MASK;
}

function splitmix64(seed: bigint): () => bigint {
  let state = toUint64(seed);
  return () => {
    state = toUint64(state + 0x9e3779b97f4a7c15n);
    let z = state;
    z = toUint64((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n);
    z = toUint64((z ^ (z >> 27n)) * 0x94d049bb133111ebn);
    return toUint64(z ^ (z >> 31n));
  };
}

function hashString(seed: string): bigint {
  let hash = 0xcbf29ce484222325n;
  for (let i = 0; i < seed.length; i += 1) {
    hash ^= BigInt(seed.charCodeAt(i));
    hash = toUint64(hash * 0x100000001b3n);
  }
  return hash;
}

function normalizeSeed(seed: SeedInput | undefined): [bigint, bigint] {
  if (typeof seed === 'number') {
    return normalizeSeed(BigInt(seed));
  }
  if (typeof seed === 'bigint') {
    const generator = splitmix64(seed);
    return [generator(), generator() || 1n];
  }
  if (typeof seed === 'string') {
    const hash = hashString(seed);
    const generator = splitmix64(hash);
    return [generator(), generator() || 1n];
  }
  if (Array.isArray(seed)) {
    let [s0, s1] = seed;
    if (typeof s0 === 'number') s0 = BigInt(s0);
    if (typeof s1 === 'number') s1 = BigInt(s1);
    s0 = toUint64(BigInt(s0));
    s1 = toUint64(BigInt(s1));
    if (s0 === 0n && s1 === 0n) {
      s1 = 1n;
    }
    return [s0, s1];
  }
  if (seed && 'algorithm' in seed) {
    if (seed.algorithm !== 'xorshift128+') {
      throw new Error(`Unsupported RNG algorithm: ${seed.algorithm}`);
    }
    const [s0, s1] = seed.state;
    let state0 = toUint64(BigInt(s0));
    let state1 = toUint64(BigInt(s1));
    if (state0 === 0n && state1 === 0n) {
      state1 = 1n;
    }
    return [state0, state1];
  }
  return normalizeSeed(1n);
}

function toHex(value: bigint): string {
  return `0x${value.toString(16).padStart(16, '0')}`;
}

export class Xorshift128Plus {
  private state0: bigint;

  private state1: bigint;

  constructor(seed?: SeedInput) {
    const [s0, s1] = normalizeSeed(seed);
    this.state0 = s0;
    this.state1 = s1;
    if (this.state0 === 0n && this.state1 === 0n) {
      this.state1 = 1n;
    }
  }

  static deserialize(serialized: SerializedRngState): Xorshift128Plus {
    return new Xorshift128Plus(serialized);
  }

  clone(): Xorshift128Plus {
    const clone = new Xorshift128Plus();
    clone.state0 = this.state0;
    clone.state1 = this.state1;
    return clone;
  }

  getState(): [bigint, bigint] {
    return [this.state0, this.state1];
  }

  serialize(): SerializedRngState {
    return {
      algorithm: 'xorshift128+',
      state: [toHex(this.state0), toHex(this.state1)],
    };
  }

  nextBigInt(): bigint {
    let s1 = this.state0;
    const s0 = this.state1;
    this.state0 = s0;
    s1 ^= (s1 << 23n) & UINT64_MASK;
    s1 ^= s1 >> 17n;
    s1 ^= s0;
    s1 ^= s0 >> 26n;
    this.state1 = toUint64(s1);
    const result = toUint64(this.state1 + s0);
    return result;
  }

  nextFloat(): number {
    const value = this.nextBigInt();
    const fraction = Number(value >> 11n) / 2 ** 53;
    return fraction;
  }

  nextInt(maxExclusive: number): number {
    if (!Number.isInteger(maxExclusive) || maxExclusive <= 0) {
      throw new RangeError('maxExclusive must be a positive integer');
    }
    const bound = BigInt(maxExclusive);
    const threshold = UINT64_MASK - (UINT64_MASK % bound);
    while (true) {
      const value = this.nextBigInt();
      if (value < threshold) {
        return Number(value % bound);
      }
    }
  }

  nextRange(minInclusive: number, maxExclusive: number): number {
    if (!Number.isInteger(minInclusive) || !Number.isInteger(maxExclusive)) {
      throw new RangeError('Range boundaries must be integers');
    }
    if (maxExclusive <= minInclusive) {
      throw new RangeError('maxExclusive must be greater than minInclusive');
    }
    const delta = maxExclusive - minInclusive;
    return minInclusive + this.nextInt(delta);
  }

  next(): number {
    return this.nextFloat();
  }
}
