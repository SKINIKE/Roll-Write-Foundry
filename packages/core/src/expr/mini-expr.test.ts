import { describe, expect, it } from 'vitest';
import {
  MiniExprEvaluationError,
  MiniExprSyntaxError,
  compileExpression,
  evaluateExpression,
  listIdentifiers,
} from './index';
import { Xorshift128Plus } from '../rng/xorshift128plus';

const baseVariables = {
  ore: 5,
  crystal: 2,
  combo: 2,
  achievements: 0,
  roll_total: 11,
  roll_high: 6,
  roll_low: 5,
  roll_1: 6,
  roll_2: 5,
  turn: 3,
};

type EvalCase = { expr: string; expected: number | boolean; precision?: number };

const deterministicCases: EvalCase[] = [
  { expr: '1 + 2 * 3', expected: 7 },
  { expr: '(1 + 2) * 3', expected: 9 },
  { expr: '4 / 2 + 1', expected: 3 },
  { expr: '10 % 3', expected: 1 },
  { expr: '2 ^ 3', expected: 8 },
  { expr: '-(3 - 5)', expected: 2 },
  { expr: '!false', expected: true },
  { expr: 'true && false', expected: false },
  { expr: 'true || false', expected: true },
  { expr: '3 > 1 && 2 < 5', expected: true },
  { expr: '3 == 3', expected: true },
  { expr: '3 != 4', expected: true },
  { expr: 'CLAMP(10, 0, 5)', expected: 5 },
  { expr: 'CLAMP(-1, 0, 5)', expected: 0 },
  { expr: 'MAX(1, 7, 3)', expected: 7 },
  { expr: 'MIN(1, 7, 3)', expected: 1 },
  { expr: 'IF(1 < 2, 10, 20)', expected: 10 },
  { expr: 'IF(false, 1, 2)', expected: 2 },
  { expr: 'ABS(-3)', expected: 3 },
  { expr: 'FLOOR(3.9)', expected: 3 },
  { expr: 'CEIL(3.1)', expected: 4 },
  { expr: 'ROUND(3.456, 2)', expected: 3.46, precision: 2 },
  { expr: 'ROUND(3.456)', expected: 3 },
  { expr: 'ore + crystal', expected: 7 },
  { expr: 'ore * 3 + crystal * 5 + combo * 4', expected: 33 },
  { expr: 'roll_total >= 10', expected: true },
  { expr: 'roll_total == roll_1 + roll_2', expected: true },
  { expr: 'roll_high >= roll_low', expected: true },
  { expr: 'ore >= 3 && crystal >= 1', expected: true },
  { expr: 'MAX(ore, crystal, combo)', expected: 5 },
  { expr: 'MIN(ore, crystal, combo)', expected: 2 },
  { expr: 'roll_total / 2', expected: 5.5, precision: 5 },
  { expr: 'ore + -crystal', expected: 3 },
  { expr: '!(ore > crystal)', expected: false },
  { expr: 'IF(roll_total >= 10, ore + 5, ore)', expected: 10 },
  { expr: 'CLAMP(roll_total, 5, 9)', expected: 9 },
  { expr: 'ROUND(roll_total / 3, 1)', expected: 3.7, precision: 1 },
  { expr: 'ABS(combo - crystal)', expected: 0 },
  { expr: 'MAX(roll_1, roll_2, ore)', expected: 6 },
  { expr: 'MIN(roll_1, roll_2, ore)', expected: 5 },
  { expr: 'ROUND(-2.5)', expected: -2 },
  { expr: 'FLOOR(-1.2)', expected: -2 },
  { expr: 'CEIL(-1.2)', expected: -1 },
  { expr: 'MAX(1, MIN(2, 3))', expected: 2 },
  { expr: 'IF(roll_total > 7 && ore < 5, ore + 2, ore)', expected: 5 },
  { expr: '((ore + crystal) * combo) - achievements', expected: 14 },
  { expr: 'roll_1 ^ 2 + roll_2 ^ 2', expected: 61 },
  { expr: 'CLAMP(roll_total - ore, 0, 10)', expected: 6 },
  { expr: 'MIN(roll_total, MAX(ore, crystal))', expected: 5 },
  { expr: 'MIN(MAX(ore, 1), roll_total)', expected: 5 },
  { expr: 'turn + ore', expected: 8 },
  { expr: 'IF(turn >= 3, 1, 0)', expected: 1 },
];

describe('MiniExpr evaluation', () => {
  it('collects identifiers from expressions', () => {
    const compiled = compileExpression('ore + MAX(crystal, 3) - turn');
    expect(Array.from(listIdentifiers(compiled.ast).values()).sort()).toEqual(['crystal', 'ore', 'turn']);
  });

  it.each(deterministicCases)(
    'evaluates %s',
    ({ expr, expected, precision }) => {
      const compiled = compileExpression(expr);
      const result = evaluateExpression(compiled, { variables: baseVariables });
      if (typeof expected === 'boolean') {
        expect(result).toBe(expected);
      } else if (precision !== undefined) {
        expect(result).toBeCloseTo(expected, precision);
      } else {
        expect(result).toBe(expected);
      }
    },
  );

  it('evaluates dice helpers deterministically with RNG', () => {
    const rng = new Xorshift128Plus(99);
    const diceCases: Array<[string, number]> = [
      ['D6(3)', 0],
      ['D(4, 2)', 0],
      ['D(8)', 0],
    ];
    const results = diceCases.map(([expr]) => {
      const compiled = compileExpression(expr);
      const value = evaluateExpression(compiled, { variables: baseVariables, rng });
      if (typeof value !== 'number') {
        throw new Error('Dice evaluation should return a number');
      }
      return value;
    });
    expect(results).toMatchSnapshot('dice-rolls');
  });

  it('rejects invalid syntax', () => {
    expect(() => compileExpression('(')).toThrow(MiniExprSyntaxError);
    expect(() => compileExpression('1 +')).toThrow(MiniExprSyntaxError);
  });

  it('rejects evaluation errors', () => {
    const compiledUnknown = compileExpression('unknownVar');
    expect(() => evaluateExpression(compiledUnknown, { variables: baseVariables })).toThrow(MiniExprEvaluationError);
    const compiledClamp = compileExpression('CLAMP(1, 2, 1)');
    expect(() => evaluateExpression(compiledClamp, { variables: baseVariables })).toThrow(MiniExprEvaluationError);
    const compiledDivZero = compileExpression('1 / 0');
    expect(() => evaluateExpression(compiledDivZero, { variables: baseVariables })).toThrow(MiniExprEvaluationError);
    const compiledDice = compileExpression('D6()');
    expect(() => evaluateExpression(compiledDice, { variables: baseVariables })).toThrow(MiniExprEvaluationError);
  });
});
