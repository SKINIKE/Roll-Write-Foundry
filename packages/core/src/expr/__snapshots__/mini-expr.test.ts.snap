// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MiniExpr evaluation > evaluates dice helpers deterministically with RNG > dice-rolls 1`] = `
[
  14,
  8,
  3,
]
`;
